### PR TITLE
Fixed `migrate-credential` cli command on AWS

### LINKED ISSUE #1726: [BUG]: AWS Migrate Credentials Doesn't work properly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

migrate-credential CLI command scans the uc_roles_access.csv file for existing UC roles and attempt creating credentials for each one of them.
It errors out trying to create the same role twice.
It doesn't create credentials for the roles created in the create-missing-principals step.

### Expected Behavior

The migrate-credential should migrate all and only the relevant roles.

### Steps To Reproduce

_No response_

### Cloud

AWS

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by FastLee**
closes #1726

### CODE REVIEW

File: src/databricks/labs/ucx/assessment/aws.py [hunk 1]
```diff
@@ -216,22 +229,33 @@ def _s3_actions(self, actions):
...
                 "Version": "2012-10-17",
-                "Statement": [self._databricks_trust_statement(external_id)],
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Principal": {
+                            "AWS": (
+                                [
+                                    "arn:aws:iam::414351767826:role/unity-catalog-prod-UCMasterRole-14S5ZJVKOTYTL",
+                                    self_assume_arn,
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:243]: can we feed `self_assume_arn` to `_databricks_trust_statement` instead?

File: src/databricks/labs/ucx/assessment/aws.py [hunk 2]
```diff
@@ -290,6 +332,15 @@ def create_uc_role(self, role_name: str) -> str | None:
         """
         return self._create_role(role_name, self._aws_role_trust_doc())
 
+    def update_uc_role(self, role_name: str, role_arn: str) -> str | None:
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:335]: what is the difference between `update_uc_role` and `update_uc_trust_role`

File: src/databricks/labs/ucx/assessment/aws.py [hunk 3]
```diff
@@ -80,6 +92,7 @@ class AWSResources:
     S3_ACTIONS: typing.ClassVar[set[str]] = {"s3:PutObject", "s3:GetObject", "s3:DeleteObject", "s3:PutObjectAcl"}
     S3_READONLY: typing.ClassVar[str] = "s3:GetObject"
     S3_REGEX: typing.ClassVar[str] = r"arn:aws:s3:::([a-zA-Z0-9\/+=,.@_-]*)\/\*$"
+    S3_BUCKET: typing.ClassVar[str] = r"((s3:\/\/|s3a:\/\/)([a-zA-Z0-9+=,.@_-]*))\/.*$"
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:95]: can we modify `S3_PATH_REGEX` instead

File: src/databricks/labs/ucx/aws/credentials.py [hunk 4]
```diff
@@ -161,16 +169,21 @@ def run(self, prompts: Prompts, include_names: set[str] | None = None) -> list[C
 
         execution_result = []
         for iam in iam_list:
-            storage_credential = self._storage_credential_manager.create(iam)
+            credential = StorageCredentialInfo(
+                name=iam.role_name,
+                aws_iam_role=AwsIamRoleResponse(role_arn=iam.role_arn),
+                read_only=iam.privilege == Privilege.READ_FILES.value,
+            )
+            storage_credential = self._storage_credential_manager.create(credential)
```
Comment (nkvuong) [inline src/databricks/labs/ucx/aws/credentials.py:177]: do we need this refactoring step?

File: src/databricks/labs/ucx/aws/access.py [hunk 5]
```diff
@@ -52,30 +55,41 @@ def list_uc_roles(self, *, single_role=True, role_name="UC_ROLE", policy_name="U
...
                 self._aws_resources.put_role_policy(
                     role.role_name,
                     role.policy_name,
-                    set(role.resource_paths),
+                    expanded_paths,
                     self._aws_account_id,
                     self._kms_key,
                 )
                 roles_created.append(role)
+        time.sleep(15)
```
Comment (nkvuong) [inline src/databricks/labs/ucx/aws/access.py:87]: can we use a `from databricks.sdk.retries import retried` instead of a fixed sleep duration

**Review by nkvuong — CHANGES_REQUESTED** [review]
the fix definitely makes sense, some suggested changes from me

File: src/databricks/labs/ucx/assessment/aws.py (hunk 3 above)
Comment (nkvuong) [reply src/databricks/labs/ucx/assessment/aws.py:95]: the other place that `S3_PATH_REGEX` is used is in `aws/access.py`, and that only checks that it matches an s3 path

Comment (FastLee) [reply src/databricks/labs/ucx/assessment/aws.py:95]: Slightly different uses. S3Path extracts the whole path. Bucket extracts just the bucket path.

File: src/databricks/labs/ucx/assessment/aws.py [hunk 6]
```diff
@@ -216,22 +230,33 @@ def _s3_actions(self, actions):
...
 
     @staticmethod
     def _databricks_trust_statement(external_id="0000"):
-        return {
-            "Effect": "Allow",
-            "Principal": {"AWS": "arn:aws:iam::414351767826:role/unity-catalog-prod-UCMasterRole-14S5ZJVKOTYTL"},
-            "Action": "sts:AssumeRole",
-            "Condition": {"StringEquals": {"sts:ExternalId": external_id}},
-        }
+        return {"StringEquals": {"sts:ExternalId": external_id}}
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:259]: why don't you make `_databricks_trust_statement` take an extra parameter of `self_assume_arn`, then the function would be like below

```suggestion
        trusted_arns = ["arn:aws:iam::414351767826:role/unity-catalog-prod-UCMasterRole-14S5ZJVKOTYTL"]
        if self_assume_arn:
            trusted_arns.append(self_assume_arn)
        return {
                      "Effect": "Allow",
                      "Principal": {
                          "AWS": trusted_arns,
                      },
                      "Action": "sts:AssumeRole",
                      "Condition": {"StringEquals": {"sts:ExternalId": external_id}},
                  }        
        }
```

File: src/databricks/labs/ucx/assessment/aws.py [hunk 7]
```diff
@@ -216,22 +230,33 @@ def _s3_actions(self, actions):
...
+                                    self_assume_arn,
+                                ]
+                                if self_assume_arn
+                                else "arn:aws:iam::414351767826:role/unity-catalog-prod-UCMasterRole-14S5ZJVKOTYTL"
+                            )
+                        },
+                        "Action": "sts:AssumeRole",
+                        "Condition": {"StringEquals": {"sts:ExternalId": external_id}},
+                    }
+                ],
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:253]: this change will then just be 
```suggestion
"Statement": [self._databricks_trust_statement(external_id, self_assume_arn)],
```

File: src/databricks/labs/ucx/assessment/aws.py [hunk 8]
```diff
@@ -246,7 +271,14 @@ def _aws_s3_policy(self, s3_prefixes, account_id, role_name, kms_key=None):
         s3_prefixes_enriched = sorted([self.S3_PREFIX + s3_prefix for s3_prefix in s3_prefixes_strip])
         statement = [
             {
-                "Action": ["s3:GetObject", "s3:PutObject", "s3:DeleteObject", "s3:ListBucket", "s3:GetBucketLocation"],
+                "Action": [
+                    "s3:GetObject",
+                    "s3:PutObject",
+                    "s3:PutObjectAcl",
```
Comment (nkvuong) [inline src/databricks/labs/ucx/assessment/aws.py:277]: our doc does not require this permission - https://docs.databricks.com/en/connect/unity-catalog/storage-credentials.html#step-1-create-an-iam-role
```suggestion
```

Comment (FastLee) [reply src/databricks/labs/ucx/assessment/aws.py:277]: It is required.

File: src/databricks/labs/ucx/assessment/aws.py (hunk 6 above)
Comment (FastLee) [reply src/databricks/labs/ucx/assessment/aws.py:259]: Why not combine it all to a single method.

File: src/databricks/labs/ucx/assessment/aws.py (hunk 7 above)
Comment (FastLee) [reply src/databricks/labs/ucx/assessment/aws.py:253]: I removed _databricks_trust_statement it is unnecessary.

File: tests/unit/assessment/test_aws.py [hunk 9]
```diff
@@ -751,3 +753,65 @@ def exists_call(cmd: str):
...
+        },
+        "Description": "",
+        "MaxSessionDuration": 3600,
+        "RoleLastUsed": {}
+    }
+}
+            """,
+                "",
+            )
+        command_calls.append(cmd)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply tests/unit/assessment/test_aws.py]: Fixed

**Review by nkvuong — APPROVED** [review]

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**